proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+Updates to the solver
+~~~~~~~~~~~~~~~~~~~~~
+Inside GHC solver each constraint would come with an optional attached message and a priority number. Whenever a new item has to be taken out of the work list, the ones with highest priority should come before. If more than one constraint have the same priority, then the choice is done as now -- first canonicalization, then interaction, and so on.
+
+Whenever the solver finds a ``IfUndischarged c msg``, it emits the constraint ``c`` with the additional message attached. This message has no influence on the rest of the solving process, but if it ends with errors and the constraint should be reported, the message is used. Indeed, we could see its definition as (note that ``UndecidableSuperClasses`` is required) ::
+
+    class c => IfUndischarged c msg
+
+Priorities and messages are inherited by constraints stemming from another one by solving. For example, if we have ``IfUndischarged (Eq [a]) msg``, we would end up with a constraint ``Eq a`` where the message ``msg`` is kept.
+
+Finally, the solver has to be updated with the following new rules for inequality checks ::
+
+    [G] IfApartUnsafe a b no yes ==> yes
+    [W] IfApartUnsafe a b no yes
+          | if a and b are apart ==> no
+          | otherwise            ==> yes
+
+Indeed, we could see `IfApartUnsafe` as defined whenever compilation is successful as (again, using ``UndecidableSuperClasses``) ::
+
+    class yes => IfApartUnsafe a b no yes
+
+
+Effect and Interactions
+-----------------------
+In principle, there should be no effect for already-existing code. Right now people do not assume any specific ordering on the constraint solver, which means that we can change it as explained here.
+
+If implemented as-is, using ``(==)`` with the aforementioned type signature would impose a performance cost. Whereas before we just needed a ``Eq`` dictionary and a type application, now we need two of them and a complex proof of the big constraint. Luckily, this is easy to fix by using an ``INLINE`` pragma ::
+
+  {-# INLINE (==) #-}
+  (==) :: CustomErrors [ ... ] => a -> b -> Bool
+  (==) = eq
+
+where ``eq`` is the function with the simpler type signature. Preliminary research shows that this solves most of the problems, but no check has been done for interaction with other pragmas or optimizations.
+
+One problematic aspect of this way to encode custom errors is that the amount of type variables in a signature tends to grow (for ``(==)`` we have already duplicated it!). This might or might not be a problem, depending on who you ask.
+
+Finally, it seems that the approach we use for solving interacts in weird ways with ambiguity checking. In our prototype built over branch 8.3, with some complex usages of ``CustomErrors`` some signatures are marked as ambiguous (and thus require ``AllowAmbiguousTypes`` which are not). This might require improvements to the ambiguity checker, maybe making it aware of ``IfUndischarged`` and ``IfApart``.
```

Comment:
Could you provide an example here? Ambiguity checking is very simple: a type `t` is ambiguous if `x, y :: t; x = undefined; y = x` fails to type-check. Perhaps there's a bug in GHC here, but this paragraph seems quite mysterious for now.